### DISCUSSION TITLE
Streams verification

### DISCUSSION THREAD

**Discussion by alisevych** [UnitTestBot/UTBotJava#1062] [Ideas]
**Description**

When a Stream is returned there should be some verification of it. 
At the moment there are only expected and actual - of Object type - without assertions: see bug #1046
Probably make deepEquals comparison of  the Objects inside a Stream.

**Idea**

Add verification of returned Stream.
When the data is generated for the input - we need to check the Stream on the output.
It's class and the objects inside.

**Use cases**

Functional interfaces returned from a method.

**Comment by alisevych** [comment]
@Damtev Please share your ideas on the topic

**Comment by Damtev** [comment]
Basically, there is a big difference between symbolic and concrete executions for Streams. So, we have corresponding symbolic wrappers for Streams that produce reasonable assemble models - for example, instantiating with `Collection#stream()` or `Arrays#stream` methods and then modifying with `map`, `filter`, and other methods of Stream. This model might be later reliably verified with the outer `deepEquals`  method in code generation.

On the other hand, the concrete execution finds the real type of Stream which is `java.util.stream.ReferencePipeline$Head` in JDK. It could not be converted to an assemble model and, since this class is inaccessible without reflection, will be created with a lot of reflection stuff in code generation and later could not be even verified with `deepEquals`. 

In summarizing, we need to find a way to produce an assemble model for Streams in concrete execution as it is in symbolic, or, at least, disable concrete execution for methods that return Streams as a result.

@sergeypospelov please share, what do you think
